Docs: Add section on using core dumps. (#33942)
envoyproxy/envoy: bazel/README.md

with `.dwp ` target. The `.dwp` file need to be presented in the same folder with the
binary for a full debugging experience.

# Debugging production core dumps

In production, you'll likely be running a version of envoy compiled with `-c
opt`. When we build envoy with `-c opt` it still has debug symbols attached
which greatly bloats the binary size.
Envoy uses fission which uses `gsplit-dwarf` to have the debug info partially in

> the `.o` but mostly in `.dwo` file for each object file.

Reviewer: When I use clang I usually get dwp files. Are you using gcc?

Reviewer 2: i was confused by that also - our debug images have `.dwp`

Author: interesting, I use clang14 to build and I just get `.dwo` in `bazel-bin` when only build envoy-static using `bazel build -c opt //source/exe:envoy-static --config=clang`

Reviewer 2: maybe its different if you build `opt`/`dbg`

Author: ah we build the debug images with `-c dbg`? If so it works for debugging `-c dbg` cores but not `-c opt` cores as the code generated is different.

Reviewer 2: it would be good if we elucidate this in these docs i think

Reviewer 2: thinking further on this ...

we build all release assets (including for Docker dbg container) using `-c opt` - its _enforced_ with a bazel transition

this is the typical product

```console
$ tar tf ~/Downloads/release.tar.zst 
dbg/
dbg/envoy
dbg/envoy-contrib
dbg/envoy-contrib.dwp
dbg/envoy.dwp
envoy
envoy-contrib
utils/
utils/su-exec
```

so its not `opt`/`dbg` i think

Reviewer 2: that said there are some differences i think in the 2 types of build

Reviewer 2: so re `.dwo` and `.dwp` i think the former are compiled together to create the latter, but its the `.dwp` files that you need for debugging

Reviewer 2: im also wondering if above sentence could be a bit more concise/user-focused - i think all we need to say is that when build with `-c opt` envoy will by default split the debug symbols out to a `.dwp` file which can be used for debugging

Most likely when you distribute the binary (or if you use the Envoy's
release docker image) you will have stripped the debug information such that

> when you list the sections of the binary it might look as follows:

Reviewer 2: how would you check that ?

Author: `size --format=SysV source/exe/envoy-static` would work, will include it in the document

protodesc_cold          612768   16612144
flags_help_cold           5827   17224912
.eh_frame_hdr          1389404   17230740
.eh_frame              7087852   18620144
.text                 43412112   25714688
.init                       27   69126800
.fini                       13   69126828
google_malloc           257110   69126848
malloc_hook                456   69383958
.plt                      6032   69384416
.tdata                     192   69394560
.tbss                     9637   69394752
.fini_array                  8   69394752
.init_array              20128   69394760
.data.rel.ro           2447856   69414896
.dynamic                   512   71862752
.got                     30880   71863264
.got.plt                  3032   71894144
.tm_clone_table              0   71901272
.data                   516960   71905280
__rseq_cs                 2624   72422240
__rseq_cs_ptr_array        656   72424864
.bss                  10255648   72425536
.comment                    97          0
Total                 82669971

```

In particular there is no debug information.

In order to get the decent debugging experience when loading a core into `gdb`
you'll need the following:

> - You should have a unstripped version of the binary e.g. compiled with

Reviewer: Should we also release dwp/dwo files separately? I started doing that in Istio to make it easier to debug.

Author: yes, we should release the dwo files and a binary that is unstripped as right now I think for production code we only release the stripped binary (for docker at least)

Author: cc @phlax  for ^. There have also been a few issues opened prior in this area

Reviewer 2: iirc its there in the debug docker images

we havent added in releases (and related parts of ci) mostly because it adds a lot of data overhead - we can probs revisit this

Author: resolving to carry on conversation in https://github.com/envoyproxy/envoy/pull/33942#discussion_r1589551013

`-c opt` that is not stripped. The code generated will be the same as the
stripped binary and thus compatible with the core generated from a stripped
binary.

> - The DWOs produced from `gsplit-dwarf` will need to be in the same relative
> directories to be picked up by GDB.

Reviewer 2: ```suggestion
- The `.dwp` produced by `gsplit-dwarf` will need to be in the same directory as the binary to be picked up by GDB.
```
im wondering if its possible to do it either way, and what are the differences

i guess first question i have is whether you definitely dont have a `.dwp` file when you compile? not sure if there is some flag requried for that

Author: I don't have a dwp when I compile by default using `bazel build -c opt //source/exe:envoy-static` maybe @kyessenov does?

From https://github.com/envoyproxy/envoy/blob/02c9e1552d9bd72850b07f9a68bbcff2549f1103/ci/build_setup.sh#L19
and https://github.com/envoyproxy/envoy/blob/02c9e1552d9bd72850b07f9a68bbcff2549f1103/ci/do_ci.sh#L212

I think CI has two targets -- `envoy-static` and `envoy-static.dwp`. 

Doing:
```
[02:04:51] kbaichoo:exe:0 $export ENVOY_BUILD_DEBUG_INFORMATION=//source/exe:envoy-static.dwp
[02:06:36] kbaichoo:exe:130 $bazel build -c opt "${ENVOY_BUILD_DEBUG_INFORMATION}"  //source/exe:envoy-static
```
I produce both of these files. Will play around with getting the DWP to work with the core later and update if the doc from the results.

Reviewer 2: so i guess the dwp file isnt required for debugging - and now i understand what you meant about relative paths to the binary

i think its probably an easier path to rebuild the dwp when devving given that you are going to have to rebuild the binary anyway, which i think is also more portable
